### PR TITLE
Feature/entry comment modal

### LINKED ISSUE #1985: Use new components in entry review modal
> NOTE: All components have been written in typescript with hooks, we'll only need to make UI changes

## Acceptance Criteria
* Change current entry comment/review modal to use new components
* No need to work a lot on styling
* Use text area for comment box and have button below the text area

### PR DESCRIPTION
**Pull request by samshara**
- Fixes #1985

## Changes

* Add entry comment modal
* Add edit entry comment
* Show comment in the list

* Changes to configurations

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [ ] permission checks
- [ ] translations

### CODE REVIEW

File: app/components/EntryCommentModal/Comment/EditCommentForm/index.tsx [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    const {
+        pending,
+        trigger: editComment,
+    } = useLazyRequest<EntryComment, FormType>({
+        url: `server://v2/entries/${comment.entry}/review-comments/${comment.id}/`,
+        method: 'PATCH',
+        body: (ctx) => ctx,
+        onSuccess: (response) => {
+            onEditSuccess(response);
+        },
```
Comment (AdityaKhatri) [inline app/components/EntryCommentModal/Comment/EditCommentForm/index.tsx]: handle form error.
Example is in login page

File: app/components/EntryCommentModal/Comment/EditCommentForm/index.tsx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        failureHeader: 'Entry comment edit',
+    });
+
+    const handleCancel = useCallback(() => {
+        onEditCancel();
+    }, [onEditCancel]);
```
Comment (AdityaKhatri) [inline app/components/EntryCommentModal/Comment/EditCommentForm/index.tsx]: Lets use onEditCancel directly

File: app/components/EntryCommentModal/Comment/index.tsx [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    comment: EntryComment;
+}
+
+const commentTypeToTextMap: { [id: number]: string } = {
+    0: 'commented',
+    1: 'verified',
+    2: 'unverified',
+    3: 'controlled',
+    4: 'uncontrolled',
+};
```
Comment (AdityaKhatri) [inline app/components/EntryCommentModal/Comment/index.tsx]: Lets keep this together with enum?

File: app/components/EntryCommentModal/Comment/index.tsx [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+function Comment(props: Props) {
+    const {
+        className,
+        comment: commentFromProps,
+    } = props;
+
+    const { user } = useContext(UserContext);
+    const [comment, setComment] = useState<EntryComment>(commentFromProps);
+    const [editMode, setEditMode] = useState<boolean>(false);
```
Comment (AdityaKhatri) [inline app/components/EntryCommentModal/Comment/index.tsx]: Lets use useModalState

File: app/components/EntryCommentModal/CommentForm/index.tsx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+const schema: FormSchema = {
+    fields: (): FormSchemaFields => ({
+        text: [requiredStringCondition],
+        commentType: [requiredCondition],
+        mentionedUsers: [requiredCondition],
+    }),
+};
+
+const defaultValue: FormType = {
+    commentType: 0,
```
Comment (AdityaKhatri) [inline app/components/EntryCommentModal/CommentForm/index.tsx]: use from enum

File: app/components/EntryCommentModal/CommentForm/index.tsx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                />
+                <ProjectMembersMultiSelectInput
+                    name="mentionedUsers"
+                    label="Assignees"
+                    value={value.mentionedUsers}
+                    projectId={projectId}
+                    onChange={setFieldValue}
+                    options={members}
+                    onOptionsChange={setMembers}
+                    error={error?.mentionedUsers?.toString()}
```
Comment (AdityaKhatri) [inline app/components/EntryCommentModal/CommentForm/index.tsx]: This should be handled separately. Look for getErrorObject I think. @tnagorra ?

File: app/components/EntryCommentModal/CommentForm/index.tsx [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        trigger: editComment,
+    } = useLazyRequest<EntryComment, FormType>({
+        url: `server://v2/entries/${entryId}/review-comments/`,
+        method: 'POST',
+        body: (ctx) => ctx,
+        onSuccess: (response) => {
+            setValue(defaultValue);
+            onSave(response);
+        },
+        failureHeader: 'Entry comment edit',
```
Comment (AdityaKhatri) [inline app/components/EntryCommentModal/CommentForm/index.tsx]: handle failure and set form error as well

File: app/components/EntryCommentModal/index.tsx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        comment,
+    }), []);
+
+    return (
+        <Modal
+            className={_cs(className, styles.entryCommentModal)}
+            heading="Entry Comments"
+            onCloseButtonClick={onModalClose}
+            bodyClassName={styles.modalBody}
+            footerActions={((commentsResponse?.count ?? 0) > maxItemsPerPage) && (
```
Comment (AdityaKhatri) [inline app/components/EntryCommentModal/index.tsx]: this check is not necessary

File: app/components/ProjectMembersSelectInput/index.tsx [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    Def,
+    'onSearchValueChange' | 'searchOptions' | 'optionsPending' | 'keySelector' | 'labelSelector' | 'totalOptionsCount' | 'onShowDropdownChange'
+> & {
+    projectId: number;
+};
+const keySelector = (d: Membership) => d.member;
+const labelSelector = (d: Membership) => d.memberName;
+
+const memberFieldQuery = {
+    fields: ['member', 'member_name'],
```
Comment (AdityaKhatri) [inline app/components/ProjectMembersSelectInput/index.tsx:21]: add limit of 50 members or so

File: app/types/entry.ts [hunk 10]
```diff
@@ -141,3 +141,52 @@ export interface EntryOptions {
     createdBy: KeyValueElement[];
     projectEntryLabel: BasicElement[];
 }
+
+
```
Comment (AdityaKhatri) [inline app/types/entry.ts]: remove whitespace

File: app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/EntryControl/EntryUncontrolCommentModal/index.tsx [hunk 11]
```diff
@@ -134,19 +123,19 @@ function EntryUncontrolCommentModal(props: Props) {
...
                 value={value.mentionedUsers}
+                projectId={projectId}
                 onChange={setFieldValue}
-                options={projectMembersResponse?.results}
-                keySelector={memberKeySelector}
-                labelSelector={memberNameSelector}
-                disabled={projectMembersPending}
+                options={members}
+                onOptionsChange={setMembers}
+                error={error?.mentionedUsers?.toString()}
```
Comment (AdityaKhatri) [inline app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/EntryControl/EntryUncontrolCommentModal/index.tsx]: Same as above

File: app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/EntryVerification/EntryUnverifyCommentModal/index.tsx [hunk 12]
```diff
@@ -145,19 +129,19 @@ function EntryUnverifyCommentModal(props: Props) {
...
                 value={value.mentionedUsers}
+                projectId={projectId}
                 onChange={setFieldValue}
-                options={projectMembersResponse?.results}
-                keySelector={memberKeySelector}
-                labelSelector={memberNameSelector}
-                disabled={projectMembersPending}
+                options={members}
+                onOptionsChange={setMembers}
+                error={error?.mentionedUsers?.toString()}
```
Comment (AdityaKhatri) [inline app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/EntryVerification/EntryUnverifyCommentModal/index.tsx]: same as above

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]
In general, all POST requests with form should have proper error handling.

File: app/components/ProjectMembersSelectInput/index.tsx (hunk 9 above)
Comment (samshara) [reply app/components/ProjectMembersSelectInput/index.tsx:21]: @AdityaKhatri then we will require pagination or search functionality

File: app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/EntryControl/EntryUncontrolCommentModal/index.tsx (hunk 11 above)
Comment (samshara) [reply app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/EntryControl/EntryUncontrolCommentModal/index.tsx]: @AdityaKhatri can you elaborate on this?

Comment (AdityaKhatri) [reply app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/EntryControl/EntryUncontrolCommentModal/index.tsx]: I am referring to the comment I made above.
The error sent here should be calculated from a helper function.

File: app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/EntryVerification/EntryUnverifyCommentModal/index.tsx (hunk 12 above)
Comment (AdityaKhatri) [reply app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/EntryVerification/EntryUnverifyCommentModal/index.tsx]: I am referring to the comment I made above.
The error sent here should be calculated from a helper function.

File: app/components/ProjectMembersSelectInput/index.tsx (hunk 9 above)
Comment (AdityaKhatri) [reply app/components/ProjectMembersSelectInput/index.tsx:21]: The search itself is a pagination method.

File: app/components/EntryCommentModal/CommentForm/index.tsx (hunk 6 above)
Comment (tnagorra) [reply app/components/EntryCommentModal/CommentForm/index.tsx]: Yes sir, look for that function. But, use getErrorString instead.

File: app/components/ProjectMembersSelectInput/index.tsx (hunk 9 above)
Comment (samshara) [reply app/components/ProjectMembersSelectInput/index.tsx:21]: @AdityaKhatri we don't have search functionality in api right now

Comment (samshara) [reply app/components/ProjectMembersSelectInput/index.tsx:21]: https://github.com/the-deep/client/issues/2077

File: app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/index.tsx [hunk 13]
```diff
@@ -112,12 +107,14 @@ function EntryItem(props: Props) {
...
-                readOnly
-            />
+            <>
+                <EntryListItem
+                    className={styles.entry}
+                    entry={entry1} // TODO remove mock entry usage appropriate
+                    framework={frameworkMockData}
+                    readOnly
+                />
+            </>
```
Comment (tnagorra) [inline app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/index.tsx]: Any reason to wrap this on a fragment?

File: app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/index.tsx [hunk 14]
```diff
@@ -83,15 +82,11 @@ function EntryItem(props: Props) {
                             >
                                 Edit Tags
                             </ButtonLikeLink>
-                            <QuickActionButton
+                            <EntryCommentModal
```
Comment (tnagorra) [inline app/views/Project/Tagging/Sources/SourcesTable/EntryList/EntryItem/index.tsx]: Is this a modal or a button?

File: app/components/ProjectMembersSelectInput/index.tsx [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    const {
+        pending,
+        response,
+    } = useRequest<MultiResponse<Membership>>(
+        {
+            url: `server://v2/projects/${projectId}/project-memberships/`,
+            method: 'GET',
+            query: memberFieldQuery,
+            failureHeader: 'Project Membership',
+        },
```
Comment (tnagorra) [inline app/components/ProjectMembersSelectInput/index.tsx]: The request should only be made when the popup is open

File: app/components/EntryVerification/index.tsx [hunk 16]
```diff
@@ -10,7 +10,7 @@ import { useModalState } from '#hooks/stateManagement';
 import { useLazyRequest } from '#base/utils/restRequest';
 import UserContext from '#base/context/UserContext';
 
-import { EntryAction } from '../constants';
+import { EntryAction } from '#components/commentConstants';
```
Comment (tnagorra) [inline app/components/EntryVerification/index.tsx:13]: Why not have this inside Comment? Do we need a "commentContacts" component?

File: app/components/EntryCommentModal/Comment/EditCommentForm/index.tsx [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+function EditCommentForm(props: Props) {
+    const {
+        className,
+        comment,
+        onEditSuccess,
+        onEditCancel,
+    } = props;
+
+    const [initialFormValue] = useState<FormType>({
+        text: comment.textHistory[0]?.text || '',
```
Comment (tnagorra) [inline app/components/EntryCommentModal/Comment/EditCommentForm/index.tsx]:
```suggestion
        text: comment.textHistory[0]?.text ?? '',
```

File: app/components/EntryCommentModal/Comment/index.tsx [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+        createdByDetails,
+        commentType,
+        mentionedUsersDetails,
+        createdAt,
+    } = comment;
+    const [latest] = textHistory;
+
+    const isEditable = useMemo(() => (
+        user?.id === createdByDetails?.id.toString() && latest
+    ), [user, createdByDetails, latest]);
```
Comment (tnagorra) [inline app/components/EntryCommentModal/Comment/index.tsx]: Can we only edit  latest text?

File: app/components/EntryCommentModal/CommentForm/index.tsx [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+                footerActions={(
+                    <>
+                        <Button
+                            name={undefined}
+                            type="submit"
+                            disabled={pristine || pending}
+                        >
+                            Comment
+                        </Button>
+                    </>
```
Comment (tnagorra) [inline app/components/EntryCommentModal/CommentForm/index.tsx]: Why wrap with fragment?

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/components/EntryCommentModal/Comment/index.tsx (hunk 18 above)
Comment (samshara) [reply app/components/EntryCommentModal/Comment/index.tsx]: yes

File: app/components/EntryVerification/index.tsx (hunk 16 above)
Comment (samshara) [reply app/components/EntryVerification/index.tsx:13]: we use those values elsewhere too.

**Review by AdityaKhatri — APPROVED** [review]

**Pull request merged**